### PR TITLE
Travis - Fix integration tests random errors

### PR DESCRIPTION
**Pull request by offerm**
Prevents travis random failures (race)
When running Travis integration tests, sometimes there are random errors:

1) Nodes fail to stop
2) A node can't execute RPC call since the server is not running yet.
3) issue #1496 

The reason for the second error is a race between LND and the test. LND opens the RPC port before the server is up. This may cause the node.start() call to complete before the LND's server is running. In some cases, there are followup RPC calls (like listpeers) which may get an error if the server is not ready.

The reason for the first problem is a race between LND startup and RPC stop call. Since the RPC port is open before the server startup is done a test may send a stop request (via RPC) before the startup is done. This may lead to a situation that the LND process will not stop.

This fix handles the problems by the following changes:
 
 1. Ensure the completion of server startup. This is done by calling DisconnectPeer with an invalid pubKey and checking the result. DisconnectPeer returns an error ("chain backend is still syncing") until the server is marked started.
 
 2. Remove the t.parallel() from TestChannelLinkBidirectionalOneHopePayments. This test is CPU intensive and creates ~966 go routings. While the test itself is OK it has impact on other tests that are running in parallel. In these other tests we can find sleep(time) that is used for sync. Due to this test the sleep(time) may be too short and may cause these tests to fail.

3. Properly set timeout (context) in tests. 

4. Temporary  fix for issue # 1496 

5. Indicates “server started” at the end of the Start() function and not at the start of it. 

As an implication of this fix, we can also remove the special handling in lnd.go for using simnet (see diff). This makes the integration tests more realistic for testnet and mainnet.

Issue #1496  is important and may cause errors in other places too.

### CODE REVIEW

**Comment by offerm** [comment]
@Roasbeef  If this PR passes all checks, it is ready to me merged.

Let me know if you have any comments.

Offer

**Comment by offerm** [comment]
@Roasbeef This is ready as far as I'm concern.

Offer

File: htlcswitch/link_test.go [hunk 1]
```diff
@@ -1686,6 +1689,9 @@ func updateState(batchTick chan time.Time, link *channelLink,
 // sleep in this test and the one below
 func TestChannelLinkBandwidthConsistency(t *testing.T) {
 	t.Parallel()
+	if !hodl.DebugBuild{
```
Comment (wpaulino) [inline htlcswitch/link_test.go]: :+1:

File: lnd.go [hunk 2]
```diff
@@ -568,45 +568,43 @@ func lndMain() error {
 		}()
 	}
 
-	// If we're not in simnet mode, We'll wait until we're fully synced to
+	// We'll wait until we're fully synced to
```
Comment (wpaulino) [inline lnd.go]: nit: expand comment to 80 chars per line.

File: lnd_test.go [hunk 3]
```diff
@@ -8759,7 +8774,7 @@ func testSwitchOfflineDelivery(net *lntest.NetworkHarness, t *harnessTest) {
 				Index: chanPoint.OutputIndex,
 			}
 
-			ctxt, _ = context.WithTimeout(ctxb, timeout)
+			ctxt, _ = context.WithTimeout(ctxb, time.Duration(time.Second * 15))
```
Comment (wpaulino) [inline lnd_test.go]: nit: `timeout` is 15 seconds so no need to modify this line.

File: lntest/harness.go [hunk 4]
```diff
@@ -238,7 +238,15 @@ func (n *NetworkHarness) TearDownAll() error {
 // current instance of the network harness. The created node is running, but
 // not yet connected to other nodes within the network.
 func (n *NetworkHarness) NewNode(name string, extraArgs []string) (*HarnessNode, error) {
-	return n.newNode(name, extraArgs, false)
+	node, err := n.newNode(name, extraArgs, false)
+
```
Comment (wpaulino) [inline lntest/harness.go]: nit: extra newline.

File: lnwire/reply_channel_range.go [hunk 5]
```diff
@@ -46,7 +46,12 @@ func (c *ReplyChannelRange) Decode(r io.Reader, pver uint32) error {
 		return err
 	}
 
-	c.EncodingType, c.ShortChanIDs, err = decodeShortChanIDs(r)
+	// special handling to avoid error deep compare
```
Comment (wpaulino) [inline lnwire/reply_channel_range.go]: This has been fixed in the latest master.

File: lntest/node.go [hunk 6]
```diff
@@ -254,6 +254,13 @@ func (hn *HarnessNode) Name() string {
 	return hn.cfg.Name
 }
 
+// SetPort can be used to change P2P port of a node
+// TODO (Offer): remove once issue 1496 is resolved
+func (hn *HarnessNode) SetPort(port int)  {
```
Comment (wpaulino) [inline lntest/node.go]: I think it would be best to wait until #1496 is resolved rather than adding this temporary hack.

File: lntest/harness.go [hunk 7]
```diff
@@ -552,7 +570,43 @@ func (n *NetworkHarness) RestartNode(node *HarnessNode, callback func() error) e
...
+	return ensureServerStarted(node)
+}
+
+// ensureServerStarted  verifies that the lnd server started.
+// It is done by using RPC calls to DisconnectPeerRequest which returns an error "chain backend is still syncing"
+// before the server started.
+// TODO (Offer): replace this hack with a field in getInfo that indicates that the server staretd
+
+func ensureServerStarted (node *HarnessNode) error{
```
Comment (wpaulino) [inline lntest/harness.go]: If `synced_with_chain` is true within the `getinfo` response, then we can most likely guarantee that the server has started.

File: htlcswitch/link_test.go [hunk 8]
```diff
@@ -251,7 +251,10 @@ func TestChannelLinkSingleHopPayment(t *testing.T) {
 // link to cope with bigger number of payment updates that commitment
 // transaction may consist.
 func TestChannelLinkBidirectionalOneHopPayments(t *testing.T) {
-	t.Parallel()
+	// TODO (offer): temporary disabled parallel execution to avoid impact on other tests that are using sleep(x) to sync excution
+	// TODO (offer): this test creates 966 parallel go routine which puts a lot of CPU pressure. Test itself is OK but impact on other tests is huge
+	// TODO (offer): The need of such high number of parallel routine should be discussed
+	//t.Parallel()
```
Comment (wpaulino) [inline htlcswitch/link_test.go]: I'm fine with removing `t.Parallel()` for this use case.

File: server.go [hunk 9]
```diff
@@ -777,6 +777,10 @@ func (s *server) Start() error {
 		srvrLog.Infof("Auto peer bootstrapping is disabled")
 	}
 
+	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
```
Comment (wpaulino) [inline server.go]: What's the reason for this change?

File: htlcswitch/switch_test.go [hunk 10]
```diff
@@ -1872,3 +1877,24 @@ func TestMultiHopPaymentForwardingEvents(t *testing.T) {
 		}
 	}
 }
+
+// TestLogTicker verifies that the logTicker ticked by waiting 11 seconds after the start of switch
+// without this the code handling the logTicker will not be called and there will be a drop in coverage stats (build fail)
+func TestLogTicker(t *testing.T) {
```
Comment (wpaulino) [inline htlcswitch/switch_test.go]: The coverage stats have been deemed a bit unreliable for some time now, so not sure if this is something we should worry about.

File: server.go (hunk 9 above)
Comment (offerm) [reply server.go]: This is needed in order to avoid a race between stop() and start() of the server (and other calls which assume that the server started when `s.started` is 1. 
The current code changes `s.started` from zero to one at the beginning of the start function. This opens the door to races.

By moving it to the end we can be sure that integration tests will not progress until the server actually completed the startup procedure.

File: lnwire/reply_channel_range.go (hunk 5 above)
Comment (offerm) [reply lnwire/reply_channel_range.go]: Nop, this was only partially fixed (for line 669).

If you add 
`numChanIDs = 0`

after line 701 at lnwire_test.go (so you force the random number to bu zero)  to make it:


```
			numChanIDs := rand.Int31n(5000)
			numChanIDs = 0

			req.ShortChanIDs = make([]ShortChannelID, numChanIDs)
```

and run the test you will see the problem.

**Comment by offerm** [comment]
@wpaulino @Roasbeef Any progress with this PR?

**Comment by halseth** [comment]
Is this PR fixing several bugs? In its current state it is very hard to review.

It should be broken up into smaller PRs that fixes one thing at a time, the commit history should be cleaned up to make it easier to review and explain what is being fixed exactly.

**Comment by offerm** [comment]
@halseth This PR solves many issues with the build process. You need all of them to be sure that Travis/Coveralls builds work without an issue.

I can break it down to several PRs but I can't be sure that Travis and Coveralls will pass when each change is standalone.

If you agree to review it even if marked as a failure, I will do it.

Alternatively, I can document all these bugs, the reason for them, and how I solved them.

Let me know, please

Offer

**Comment by offerm** [comment]
@halseth ?

**Comment by halseth** [comment]
@offerm We should definitely make each change standalone. If they are all fixing existing errors in master, it shouldn't make the build more likely to fail.

**Comment by offerm** [comment]
@halseth Created 8 different PRs for these changes so it will be super easy to understand.
These are #1591 #1592 #1593 #1594 #1590 #1589 #1588 #1587 

As you can see almost all these PRs had travis errors which are not related at all to the change done in the the PR that failed and are fixed by one of the other PRs.

Priority (based on likelihood of showing up):
1: #1588 #1589 #1590 #1591 
2: #1594 
3: #1592 
4: #1587 #1593

**Comment by halseth** [comment]
@offerm Thanks! Should be much easier to review now :)

Closing, and moving the discussion to the relevant PRs.

**Comment by offerm** [comment]
@halseth are these 8 PRs going to be reviewed?  So far only one of them got attention

**Comment by halseth** [comment]
They will be reviewed eventually :) The complex ones will take longer to review, obviouslyl
